openfe/protocols/openmm_septop/base.py

```diff
@@ -0,0 +1,1300 @@
+                    atom_indices.extend([atom.index for atom in residue.atoms()])
+            comp_atomids[key] = atom_indices
+        return comp_atomids
+
+    @staticmethod
+    def get_smc_comps(
+        alchem_comps: dict[str, list[Component]],
+        smc_comps: dict[SmallMoleculeComponent, OFFMolecule],
+    ) -> tuple[dict[SmallMoleculeComponent, OFFMolecule],
+               dict[SmallMoleculeComponent, OFFMolecule],
+               dict[SmallMoleculeComponent, OFFMolecule],
+               dict[SmallMoleculeComponent, OFFMolecule]]:
+        # Get smcs for the different states and the common smcs
+        smc_off_A = {m: m.to_openff() for m in alchem_comps['stateA']}
+        smc_off_B = {m: m.to_openff() for m in alchem_comps['stateB']}
+        # Common smcs could e.g. be cofactors
+        smc_off_both = {m: m.to_openff() for m in smc_comps
+                        if (m not in alchem_comps["stateA"] and m not in
+                            alchem_comps["stateB"])}
+        smc_comps_A = smc_off_A | smc_off_both
+        smc_comps_B = smc_off_B | smc_off_both
+        smc_comps_AB = smc_off_A | smc_off_B | smc_off_both
+
+        return smc_comps_A, smc_comps_B, smc_comps_AB, smc_off_B
+
+    def get_system(
+        self,
+        solv_comp: SolventComponent,
+        prot_comp: ProteinComponent,
+        smc_comp: dict[SmallMoleculeComponent, OFFMolecule],
+        settings: dict[str, SettingsBaseModel],
+    ):
+        """
+        Creates an OpenMM system, topology, positions, modeller and also
+        residue IDs of the different components
+
+        Parameters
+        ----------
+        solv_comp: SolventComponent
+        prot_comp: Optional[ProteinComponent]
+        smc_comp: dict[SmallMoleculeComponent,OFFMolecule]
+        settings: dict[str, SettingsBaseModel]
+          A dictionary of settings object for the unit.
+
+        Returns
+        -------
+        omm_system: app.System
+        omm_topology: app.Topology
+        positions: openmm.unit.Quantity
+        system_modeller: app.Modeller
+        comp_resids: dict[Component, npt.NDArray]
+          A dictionary of residues for each component in the System.
+        """
+        # Get system generator
+        system_generator = self._get_system_generator(settings, solv_comp)
+
+        # Get modeller
+        system_modeller, comp_resids = self._get_modeller(
+            prot_comp, solv_comp, smc_comp,
+            system_generator, settings['solvation_settings'],
+        )
+
+        # Get OpenMM topology, positions and system
+        omm_topology, omm_system, positions = self._get_omm_objects(
+            system_modeller, system_generator, list(smc_comp.values())
+        )
+
+        return omm_system, omm_topology, positions, system_modeller, comp_resids
+
+
+class BaseSepTopRunUnit(gufe.ProtocolUnit):
+    """
+    Base class for running ligand SepTop RBFE free energy transformations.
+    """
+
+    def _prepare(
```

Comment:
This looks to be the same function in the `BaseSepTopSetupUnit` could we use a mixin?

Reply:
Might not be worth overly optimizing on this one - in "the refactor" they will both inherit this from some base unit that implements prepare (since it's really needed everywhere).